tests/test_hll.erl

```diff
@@ -0,0 +1,179 @@
+    check_precision_and_reduce_invalid_test(C, CMod, Bucket, ?DEFAULT_P - 1,
+                                       HllSet1),
+
+    ok.
+
+pb_tests(C, CMod, Bucket, Nodes) ->
+    lager:info("PB CLI TESTS: Create new Hll DT"),
+    S0 = riakc_hll:new(),
+
+    add_tests(C, CMod, Bucket, S0),
+
+    HllSet0 = get_hll(C, CMod, Bucket),
+    check_precision_and_reduce_test(C, CMod, Bucket, ?SET_P, HllSet0),
+
+    partition_write_heal(C, CMod, Bucket, Nodes),
+
+    HllSet1 = get_hll(C, CMod, Bucket),
+    check_precision_and_reduce_invalid_test(C, CMod, Bucket, ?SET_P - 1, HllSet1),
+
+    ok.
+
+add_tests(C, CMod, Bucket, NewS) ->
+    add_element(C, CMod, Bucket, NewS, <<"OH">>),
+    {ok, S1} = CMod:fetch_type(C, Bucket, ?KEY),
+    ?assertEqual(riakc_hll:value(S1), 1),
+
+    add_elements(C, CMod, Bucket, S1, [<<"C">>, <<"A">>, <<"P">>]),
+    {ok, S2} = CMod:fetch_type(C, Bucket, ?KEY),
+    ?assertEqual(riakc_hll:value(S2), 4),
+
+    add_redundant_element(C, CMod, Bucket, S2, <<"OH">>),
+    {ok, S3} = CMod:fetch_type(C, Bucket, ?KEY),
+    ?assertEqual(riakc_hll:value(S3), 4).
+
+partition_write_heal(C, CMod, Bucket, Nodes) ->
+    lager:info("Partition cluster in two to force merge."),
+    [N1, N2, N3, N4] = Nodes,
+    PartInfo = rt:partition([N1, N2], [N3, N4]),
+
+    lager:info("Write to one side of the partition"),
+    {ok, S0} = CMod:fetch_type(C, Bucket, ?KEY),
+    add_element(C, CMod, Bucket, S0, <<"OH hello there">>),
+    {ok, S1} = CMod:fetch_type(C, Bucket, ?KEY),
+    ?assertEqual(riakc_hll:value(S1), 5),
+
+    lager:info("Heal"),
+    ok = rt:heal(PartInfo),
+    ok = rt:wait_for_cluster_service(Nodes, riak_kv).
+
+get_hll(C, CMod, Bucket) ->
+    {ok, Obj} =CMod:get(C, Bucket, ?KEY),
+    {ok, CRDT} = riak_kv_crdt:from_binary(riakc_obj:get_value(Obj)),
+    {_, _, _, HllSet} = CRDT,
+    HllSet.
+
+add_element(C, CMod, Bucket, S, Elem) ->
+    lager:info("Add element to HLL DT"),
+    CMod:update_type(
+      C, Bucket, ?KEY,
+      riakc_hll:to_op(
+        riakc_hll:add_element(Elem, S))).
+
+add_elements(C, CMod, Bucket, S, Elems) ->
+    lager:info("Add multiple elements to HLL DT"),
+    CMod:update_type(
+      C, Bucket, ?KEY,
+      riakc_hll:to_op(
+        riakc_hll:add_elements(Elems, S))).
+
+add_redundant_element(C, CMod, Bucket, S, Elem) ->
+    lager:info("Add redundant element to HLL DT by calling"
+               " add_element/3 again"),
+    add_element(C, CMod, Bucket, S, Elem).
+
+check_precision_and_reduce_test(C, CMod, Bucket, ExpP, HllSet) ->
+    {ok, Props0} = CMod:get_bucket(C, Bucket),
+    ?assertEqual(proplists:get_value(?P_SETTING, Props0), ExpP),
+    ?assertEqual(riak_dt_hll:precision(HllSet), ExpP),
+    ok = CMod:set_bucket(C, Bucket, [{?P_SETTING, ExpP - 1}]),
+    {ok, Props1} = CMod:get_bucket(C, Bucket),
+    ?assertEqual(proplists:get_value(?P_SETTING, Props1), ExpP - 1).
+
+check_precision_and_reduce_invalid_test(C, CMod, Bucket, ExpP, HllSet) ->
```

Comment:
Is this because HLL's can have their precision reduced but not increased? If so please add a comment that makes that explicit for the less educated reader.